Improve solutions/analysis for Mountain View and Rental Service (#3646)
cpinitiative/usaco-guide: solutions/silver/usaco-896.mdx

![](usaco-896/mountains0.png)

Therefore, for a mountain $\mathcal{i}$ to be contained in another mountain $\mathcal{j}$, $\mathcal{i}_{x}+{i}_{y}\le{j}_{x}+{j}_{y}$ and $\mathcal{i}_{x}-{i}_{y}\ge{j}_{x}-{j}_{y}$.

We can see this concept in the graphs below:

![](usaco-896/mountains1.png)

In this case, In this case, $\mathcal{j}$, $\mathcal{i}_{x}+{i}_{y}\le{j}_{x}+{j}_{y}$ and $\mathcal{i}_{x}-{i}_{y}\ge{j}_{x}-{j}_{y}$, meaning  that the left and rightmost end of the mountain $\mathcal{i}$ are contained within mountain $\mathcal{j}$.

![](usaco-896/mountains2.png)

In this case, $\mathcal{j}$, $\mathcal{i}_{x}+{i}_{y}\le{j}_{x}+{j}_{y}$, but $\mathcal{i}_{x}-{i}_{y}<{j}_{x}-{j}_{y}$, meaning  that the  rightmost end of the mountain $\mathcal{i}$ is to the left of the rightmost end of mountain $\mathcal{j}$, but the leftmost end of mountain $\mathcal{i}$ is not to the right of the leftmost end of mountain $\mathcal{j}$, meaning it is not contained within mountain j.

This same reasoning can also work if the mountains overlap, as shown below:

![](usaco-896/mountains3.png)

> Using this information, we can loop through every pair of mountains and check if one is occluded by another, and complete this problem in $\mathcal{O}(n^2)$ time complexity, but that may be too slow and not complete every testcase (our Java implementation of this method completed 10/11 testcases), so we can use an alternative approach.

Reviewer: just say "naively going through all pairs is too slow lol"

![](usaco-896/mountains0.png)

Therefore, for a mountain $\mathcal{i}$ to be contained in another mountain $\mathcal{j}$, $\mathcal{i}_{x}+{i}_{y}\le{j}_{x}+{j}_{y}$ and $\mathcal{i}_{x}-{i}_{y}\ge{j}_{x}-{j}_{y}$.

We can see this in the graphs below:

![](usaco-896/mountains1.png)

In this case, In this case, $\mathcal{j}$, $\mathcal{i}_{x}+{i}_{y}\le{j}_{x}+{j}_{y}$ and $\mathcal{i}_{x}-{i}_{y}\ge{j}_{x}-{j}_{y}$, meaning that the left and right ends of the mountain $\mathcal{i}$ are contained within mountain $\mathcal{j}$.

![](usaco-896/mountains2.png)

In *this* case, $\mathcal{j}$, $\mathcal{i}_{x}+{i}_{y}\le{j}_{x}+{j}_{y}$, but $\mathcal{i}_{x}-{i}_{y}<{j}_{x}-{j}_{y}$, meaning that the right end of the mountain $\mathcal{i}$ is to the left of the rightmost end of mountain $\mathcal{j}$, but the left end of mountain $\mathcal{i}$ is not to the right of the left end of mountain $\mathcal{j}$, meaning it is not contained within mountain j.

This same reasoning can also work if the mountains overlap, as shown below:

> ![](usaco-896/mountains3.png)

Reviewer: ok i'm all for handholding but this seems a bit _too_ obvious

i don't think it should take this many lines to say "a mountain is hidden if it's 100% in the interval of another"

Reviewer 2: ^^ this editorial is a tad bit too long and might confuse the reader more.

Using this information, we can loop through every pair of mountains and check if one is occluded by another, and complete this problem in $\mathcal{O}(n^2)$ time complexity, but that may be too slow and not complete every testcase (our Java implementation of this method completed 10/11 testcases), so we can use an alternative approach.

> If we sort the mountains by their left end.
> Then we can loop through the list and keep track of the highest right end.

Reviewer: ```suggestion
Let's sort the mountains by their left end first and then iterate through all of them, keeping track of the highest right end we've seen.
```
